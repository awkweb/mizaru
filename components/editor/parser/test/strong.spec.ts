import Parser from '../'

describe('basic', () => {
    for (const syntax of ['**', '__']) {
        const doc = `${syntax}foo${syntax}`
        test(doc, () => {
            const { decorations, nodes } = Parser.parse(doc)
            expect(nodes).toEqual([
                {
                    from: 0,
                    to: 9,
                    type: 'paragraph',
                    marks: [{ from: 1, to: 8, type: 'strong' }],
                },
            ])
            expect(decorations).toEqual([
                { from: 1, to: 3, type: 'syntax' },
                { from: 6, to: 8, type: 'syntax' },
            ])
        })
    }
})

describe('with nested', () => {
    test('delete', () => {
        const doc = '**~~foo~~**'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 13,
                type: 'paragraph',
                marks: [
                    { from: 3, to: 10, type: 'delete' },
                    { from: 1, to: 12, type: 'strong' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 3, to: 5, type: 'syntax' },
            { from: 8, to: 10, type: 'syntax' },
            { from: 1, to: 3, type: 'syntax' },
            { from: 10, to: 12, type: 'syntax' },
        ])
    })

    test('emphasis', () => {
        const doc = '***foo***'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 11,
                type: 'paragraph',
                marks: [
                    { from: 3, to: 8, type: 'emphasis' },
                    { from: 1, to: 10, type: 'strong' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 3, to: 4, type: 'syntax' },
            { from: 7, to: 8, type: 'syntax' },
            { from: 1, to: 3, type: 'syntax' },
            { from: 8, to: 10, type: 'syntax' },
        ])
    })

    test('inlineCode', () => {
        const doc = '**`foo`**'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 11,
                type: 'paragraph',
                marks: [
                    { from: 3, to: 8, type: 'inlineCode' },
                    { from: 1, to: 10, type: 'strong' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 3, to: 4, type: 'syntax' },
            { from: 7, to: 8, type: 'syntax' },
            { from: 1, to: 3, type: 'syntax' },
            { from: 8, to: 10, type: 'syntax' },
        ])
    })
})
