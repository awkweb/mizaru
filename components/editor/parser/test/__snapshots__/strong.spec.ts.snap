// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`** backslash 1`] = `
Object {
  "counter": 18,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 17,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 17,
          "type": "strong",
        },
      ],
      "to": 18,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** escaped 1`] = `
Object {
  "counter": 10,
  "decorations": Array [],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [],
      "to": 10,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** multiple words 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 14,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 16,
          "type": "strong",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** single word 1`] = `
Object {
  "counter": 9,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 8,
          "type": "strong",
        },
      ],
      "to": 9,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** split across multiple lines 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** with nested delete 1`] = `
Object {
  "counter": 21,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 9,
      "type": "syntax",
    },
    Object {
      "from": 12,
      "to": 14,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 18,
      "to": 20,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 14,
          "type": "delete",
        },
        Object {
          "from": 1,
          "to": 20,
          "type": "strong",
        },
      ],
      "to": 21,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** with nested emphasis 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 12,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 12,
          "type": "emphasis",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** with nested inlineCode 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 12,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 12,
          "type": "inlineCode",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`** with nested strong 1`] = `
Object {
  "counter": 21,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 7,
      "to": 9,
      "type": "syntax",
    },
    Object {
      "from": 12,
      "to": 14,
      "type": "syntax",
    },
    Object {
      "from": 18,
      "to": 20,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 9,
          "type": "strong",
        },
        Object {
          "from": 12,
          "to": 20,
          "type": "strong",
        },
      ],
      "to": 21,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ backslash 1`] = `
Object {
  "counter": 18,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 17,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 17,
          "type": "strong",
        },
      ],
      "to": 18,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ escaped 1`] = `
Object {
  "counter": 10,
  "decorations": Array [],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [],
      "to": 10,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ multiple words 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 14,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 16,
          "type": "strong",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ single word 1`] = `
Object {
  "counter": 9,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 8,
          "type": "strong",
        },
      ],
      "to": 9,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ split across multiple lines 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ with nested delete 1`] = `
Object {
  "counter": 21,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 9,
      "type": "syntax",
    },
    Object {
      "from": 12,
      "to": 14,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 18,
      "to": 20,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 14,
          "type": "delete",
        },
        Object {
          "from": 1,
          "to": 20,
          "type": "strong",
        },
      ],
      "to": 21,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ with nested emphasis 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 12,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 12,
          "type": "emphasis",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ with nested inlineCode 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 12,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 16,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 12,
          "type": "inlineCode",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "strong",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`__ with nested strong 1`] = `
Object {
  "counter": 21,
  "decorations": Array [
    Object {
      "from": 7,
      "to": 9,
      "type": "syntax",
    },
    Object {
      "from": 12,
      "to": 14,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 3,
      "type": "syntax",
    },
    Object {
      "from": 18,
      "to": 20,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 7,
          "to": 14,
          "type": "strong",
        },
        Object {
          "from": 1,
          "to": 20,
          "type": "strong",
        },
      ],
      "to": 21,
      "type": "paragraph",
    },
  ],
}
`;
